fairscale/nn/misc/flatten_params_wrapper.py

```diff
@@ -86,19 +164,15 @@ def _init_flatten_params(self) -> List[Tensor]:
                         shared_param_memo[p] = (m, n)
                         param_infos.append((m, n))
                         param_full_infos.append((module_name, n))
-                        params.append(p.detach())
-                        param_numels.append(p.numel())
-                        param_shapes.append(p.size())
+                        params.append(p)
         del shared_param_memo
 
-        assert len(set(p.dtype for p in params)) <= 1, "expects all parameters in module to have same dtype"
+        assert len(set(p.dtype for p in params)) == 1, "expects all parameters in module to have same dtype"
```

Comment:
Is empty module impossible?

Reply:
ah, very good catch. I changed it back. Added a TODO to check the handling of empty module too.